src/sandstorm/run-bundle.c++

```diff
@@ -1556,36 +1563,65 @@ private:
+      args.add("/bin/mongod");
+      args.add("--fork");
+      args.add("--bind_ip");
+      args.add("127.0.0.1");
+      args.add("--port");
+      args.add(kj::str(config.mongoPort).cStr());
+      args.add("--dbpath");
+      args.add("/var/mongo");
+      args.add("--logpath");
+      args.add("/var/log/mongo.log");
+      args.add("--pidfilepath");
+      args.add("/var/pid/mongo.pid");
+      args.add("--auth");
+      args.add("--nohttpinterface");
+      args.add("--noprealloc");
+      args.add("--nopreallocj");
+      args.add("--smallfiles");
+      args.add("--replSet");
+      args.add("ssrs");
+      args.add("--oplogSize");
+      args.add("16");
+
+      if (doMongoRepair) {
+        args.add("--repair");
+      }
+
+      args.add(nullptr);
+
+      context.warning(kj::str("MONGO ARGS: ", args));
+
+      KJ_SYSCALL(execv(args[0], const_cast<char**>(args.begin())));
       KJ_UNREACHABLE;
     });
 
     // Wait for mongod to return, meaning the database is up.  Then get its real pid via the
     // pidfile.
-    process.waitForSuccess();
-
-    // Even after the startup command exits, MongoDB takes exactly two seconds to elect itself as
-    // master of the repl set (of which it is the only damned member). Unforutnately, if Node
-    // connects during this time, it fails, sometimes without actually exiting, leaving the entire
-    // server hosed. It appears that this always takes exactly two seconds from startup, since
-    // MongoDB does some sort of heartbeat every second where it checks the replset status, and it
-    // takes three of these for the election to complete, and the first of the three happens
-    // immediately on startup, meaning the last one is two seconds in. So, we'll sleep for 3
-    // seconds to be safe.
-    // TODO(cleanup): There must be a better way...
-    int n = 3;
-    while (n > 0) n = sleep(n);
+    int exitCode = process.waitForExit();
+
+    if (exitCode == 0) {
+      // Even after the startup command exits, MongoDB takes exactly two seconds to elect itself as
+      // master of the repl set (of which it is the only damned member). Unforutnately, if Node
+      // connects during this time, it fails, sometimes without actually exiting, leaving the entire
+      // server hosed. It appears that this always takes exactly two seconds from startup, since
+      // MongoDB does some sort of heartbeat every second where it checks the replset status, and it
+      // takes three of these for the election to complete, and the first of the three happens
+      // immediately on startup, meaning the last one is two seconds in. So, we'll sleep for 3
+      // seconds to be safe.
+      // TODO(cleanup): There must be a better way...
+      int n = 3;
+      while (n > 0) n = sleep(n);
+    } else {
+      context.warning(kj::str("** Mongo exited with exitCode: ", exitCode));
+    }
 
     return KJ_ASSERT_NONNULL(parseUInt(trim(readAll("/var/pid/mongo.pid")), 10));
```

Comment:
The reason you're running into trouble with waitpid() and kill() later is because this line is returning a bogus PID in the repair case, since no Mongo daemon is started in that case. Probably it's returning the PID from the previous run.

Note that before your change, startMongo() always threw an exception if Mongo failed to start, but now you've changed the code to print a warning on error rather than throw.

Fundamentally, the problem here is that the "calling convention" of mongod is completely different when you use --repair. Normally, the process we exec forks a daemon and exits quickly, hence we wait for it to exit before continuing, and then we monitor the daemon process. But with --repair there is no daemon; the process we need to "monitor" is the one *we* forked. So perhaps what you want to do is, in the repair case, don't call waitForExit() at all, but just return the PID of the direct child process.

However, I think fundamentally the whole approach here -- of trying to use the same code to launch mongod and repair mongod -- is probably too hacky. When we want to do a repair, we should probably do it synchronously.